segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/TableIterator.java

```diff
@@ -0,0 +1,326 @@
+                    // below it. (this is very similar to the AttributeMixer - maybe reuse that methodology).
+                    boolean overridden = false;
+                    while (!this.cacheHashes.isEmpty()) {
+                        val cacheHash = this.cacheHashes.peekFirst();
+                        int cmp = indexHash.getKey().compareTo(this.cacheHashes.peekFirst().getKey());
+                        if (cmp < 0) {
+                            // Cache Hash is after Index Hash. We are done here.
+                            break;
+                        }
+
+                        // The value we got from the Index was overridden by the one in the cache.
+                        overridden = overridden || cmp == 0;
+                        add(cacheHash, buckets);
+                        this.cacheHashes.removeFirst();
+                    }
+
+                    if (!overridden) {
+                        add(indexHash, buckets);
+                    }
+                }
+            }
+        }
+
+        return buckets;
+    }
+
+    private void add(Map.Entry<UUID, Long> bucketInfo, List<TableBucket> buckets) {
+        buckets.add(new TableBucket(bucketInfo.getKey(), bucketInfo.getValue()));
+    }
+
+    //endregion
+
+    //region Builder
+
+    /**
+     * Creates a new {@link TableIterator.Builder} that can be used to construct {@link TableIterator} instances.
+     *
+     * @param <T> Type of the elements returned at each iteration.
+     * @return A new instance of the {@link TableIterator.Builder} class.
+     */
+    static <T> Builder<T> builder() {
+        return new Builder<>();
+    }
+
+    /**
+     * Creates a new {@link TableIterator} that contains no elements.
+     *
+     * @param <T> Type of elements returned at each iteration.
+     * @return A new instance of the {@link TableIterator.Builder} class.
+     */
+    static <T> TableIterator<T> empty() {
+        return new TableIterator<>(
+                () -> CompletableFuture.completedFuture(null),
+                ignored -> CompletableFuture.completedFuture(null),
+                new ArrayDeque<>(),
+                ForkJoinPool.commonPool());
+    }
+
+    /**
+     * Builder for the {@link TableIterator} class.
+     */
+    static class Builder<T> {
+        private DirectSegmentAccess segment;
+        private Map<UUID, CacheBucketOffset> cacheHashes;
+        private UUID firstHash;
+        private ConvertResult<T> resultConverter;
+        private ScheduledExecutorService executor;
+        private Duration fetchTimeout;
+
+        /**
+         * Sets a {@link DirectSegmentAccess} representing a Table Segment that the iterator will iterate over.
+         *
+         * @param segment The {@link DirectSegmentAccess} to associate.
+         * @return This object.
+         */
+        Builder<T> segment(@NonNull DirectSegmentAccess segment) {
```

Comment:
If you want to follow some of the practices we have been following in other builders across the project, I suggest you use:

```
forSegment()
withCacheHashes()
withFirstHash()
<the other ones are all with*>
```

Reply:
I am following Lombok's builder naming scheme: each field that requires a value to be set gets a method with that exact field's name, and a single argument with the value to set.

I also anticipate a question like "why can't we just use Lombok's builder then?". My answer to that would be that Lombok's builder does not build generic types well and I need a different set of arguments into the Builder, which does some processing, and then invokes the private constructor with the result of that processing.